src/impl/dense.jl

```diff
@@ -85,13 +85,128 @@ function CRC.rrule(
     𝒫weight, 𝒫x, 𝒫b = CRC.ProjectTo(weight), CRC.ProjectTo(x), CRC.ProjectTo(b)
 
     ∇fused_dense = @closure Δ -> begin
-        ∂y = ∇activation(CRC.unthunk(Δ), z, gelu, y)
+        ∂y = ∇activation(CRC.unthunk(Δ), z, NNlib.gelu, y)
         ∂w, ∂x, ∂b = ∇matmul_bias(∂y, weight, x, b)
         return ∂∅, ∂∅, ∂∅, 𝒫weight(∂w), 𝒫x(∂x), 𝒫b(∂b)
     end
 
     return z, ∇fused_dense
 end
 
+# TODO: We can optimize these a bit further by checking for cases where the forward pass
+#       is not needed. We skip such optimizations for now
+function EnzymeRules.augmented_primal(cfg, ::EnzymeCore.Const{typeof(fused_dense!)},
+        ::Type{EnzymeCore.Const{Nothing}}, y::EnzymeCore.Annotation{<:AbstractMatrix},
+        opmode::EnzymeCore.Const{<:AbstractInternalArrayOpMode}, act::EnzymeCore.Const,
+        weight::EnzymeCore.Annotation{<:AbstractMatrix},
+        x::EnzymeCore.Annotation{<:AbstractMatrix},
+        b::EnzymeCore.Annotation{<:Optional{<:AbstractVector}})
+
+    # NOTE: Here we are using the ChainRulesCore rrules if they are defined for simplicity
+    all_const = weight isa EnzymeCore.Const && b isa EnzymeCore.Const &&
+                x isa EnzymeCore.Const
+    intermediate_not_needed = unsafe_known(activation_intermediate_not_needed(
+        act.val, eltype(y.val))) || all_const
+
+    weight_cache = EnzymeRules.overwritten(cfg)[5] && !(x isa EnzymeCore.Const) &&
+                   !(y isa EnzymeCore.Const) ? copy(weight.val) : nothing
+    x_cache = EnzymeRules.overwritten(cfg)[6] && !(weight isa EnzymeCore.Const) &&
+              !(y isa EnzymeCore.Const) ? copy(x.val) : nothing
+
+    case_specific_cache = if act.val === NNlib.gelu &&
+                             opmode.val isa GPUBroadcastOp{CUDADevice}
+        tmp = similar(y.val)
+        cublasLt_fused_dense!(y.val, act, weight.val, x.val, b.val, tmp)
+        (1, tmp)
+    elseif intermediate_not_needed
+        fused_dense!(y.val, opmode.val, act.val, weight.val, x.val, b.val)
+        (1, NotaNumber())
+    elseif unsafe_known(activation_has_rrule(act.val, eltype(y.val)))
+        tmp = matmuladd(weight.val, x.val, b.val)
+        activation!(y.val, opmode.val, act.val, tmp)
+        (1, tmp)
+    else
+        # TODO: Here for performance we might want to fuse the bias and activation together.
+        #       We skip this optimization for now
+        matmuladd!(y.val, opmode.val, weight.val, x.val, b.val)
+        tmp = zero.(y.val)
+        EnzymeCore.autodiff(EnzymeCore.Forward, EnzymeCore.Const(activation!),
+            EnzymeCore.Duplicated(y.val, tmp), opmode, act,
+            EnzymeCore.Duplicated(y.val, one.(y.val)))
+        (2, tmp)
```

Comment:
@wsmoses can you verify that this won't lead to strange issues? Essentially I am computing the broacasted activation gradient using forward mode. On CPU everything works correctly, but on GPU I am getting:

```julia
Call parameter type does not match function signature!
i8* getelementptr inbounds ([303 x i8], [303 x i8]* @enz_exception, i32 0, i32 0)
 i64  call void @gpu_report_exception(i8* getelementptr inbounds ([303 x i8], [303 x i8]* @enz_exception, i32 0, i32 0)), !dbg !403
; Function Attrs: mustprogress willreturn
define internal fastcc "enzyme_type"="{[-1]:Float@float}" float @preprocess_julia_sigmoid_fast_358614(float "enzyme_type"="{[-1]:Float@float}" "enzymejl_parmtype"="134156788071136" "enzymejl_parmtype_ref"="0" %0) unnamed_addr #11 !dbg !376 {
top:
  %1 = call {}*** @julia.get_pgcstack() #12
  %2 = call float @__nv_fabsf(float %0) #12, !dbg !377
  %3 = fneg fast float %2, !dbg !380
  %4 = call float @__nv_fast_expf(float %3) #12, !dbg !381
  %5 = fcmp ult float %0, 0.000000e+00, !dbg !382
  %6 = fadd float %4, 1.000000e+00, !dbg !386
  %7 = call float @__nv_frcp_rn(float %6) #13, !dbg !388
  %8 = fdiv float %4, %6, !dbg !389
  %9 = select i1 %5, float %8, float %7, !dbg !390
  %10 = fcmp ule float %0, 4.000000e+01, !dbg !391
  %11 = fcmp uge float %0, -8.000000e+01, !dbg !395
  %12 = select i1 %11, float %9, float 0.000000e+00, !dbg !397
  %13 = select i1 %10, float %12, float 1.000000e+00, !dbg !397
  ret float %13, !dbg !397
}

; Function Attrs: mustprogress willreturn
define internal fastcc { float, float } @fwddiffejulia_sigmoid_fast_358614(float "enzyme_type"="{[-1]:Float@float}" "enzymejl_parmtype"="134156788071136" "enzymejl_parmtype_ref"="0" %0, float "enzyme_type"="{[-1]:Float@float}" "enzymejl_parmtype"="134156788071136" "enzymejl_parmtype_ref"="0" %"'") unnamed_addr #11 !dbg !398 {
top:
  %1 = call {}*** @julia.get_pgcstack() #12
  %2 = call float @__nv_fabsf(float %0) #12, !dbg !399
  %3 = fcmp fast olt float %0, 0.000000e+00, !dbg !402
  %4 = select fast i1 %3, float -1.000000e+00, float 1.000000e+00, !dbg !402
  %5 = fmul fast float %"'", %4, !dbg !402
  %6 = fneg fast float %2, !dbg !402
  %7 = fneg fast float %5, !dbg !403
  call void @gpu_report_exception(i8* getelementptr inbounds ([303 x i8], [303 x i8]* @enz_exception, i32 0, i32 0)), !dbg !403
  call void @gpu_signal_exception(), !dbg !403
  call void @llvm.trap() #13, !dbg !403
  %8 = call { float, float } @fwddiffe__nv_fast_expf(float %6, float %7), !dbg !403
  %9 = extractvalue { float, float } %8, 0, !dbg !404
  %10 = extractvalue { float, float } %8, 1, !dbg !404
  %11 = fcmp ult float %0, 0.000000e+00, !dbg !404
  %12 = fadd float %9, 1.000000e+00, !dbg !408
  %13 = call float @__nv_frcp_rn(float %12) #14, !dbg !410
  %14 = fmul fast float %12, %12, !dbg !411
  %15 = fdiv fast float %10, %14, !dbg !411
  %16 = fneg fast float %15, !dbg !411
  %17 = fdiv float %9, %12, !dbg !411
  %18 = fmul fast float %10, %12, !dbg !412
  %19 = fmul fast float %10, %9, !dbg !412
  %20 = fsub fast float %18, %19, !dbg !412
  %21 = fmul fast float %12, %12, !dbg !412
  %22 = fdiv fast float %20, %21, !dbg !412
  %"'ipse" = select fast i1 %11, float %22, float %16, !dbg !412
  %23 = select i1 %11, float %17, float %13, !dbg !412
  %24 = fcmp ule float %0, 4.000000e+01, !dbg !413
  %25 = fcmp uge float %0, -8.000000e+01, !dbg !417
  %"'ipse9" = select fast i1 %25, float %"'ipse", float 0.000000e+00, !dbg !419
  %26 = select i1 %25, float %23, float 0.000000e+00, !dbg !419
  %"'ipse10" = select fast i1 %24, float %"'ipse9", float 0.000000e+00, !dbg !419
  %27 = select i1 %24, float %26, float 1.000000e+00, !dbg !419
  %28 = insertvalue { float, float } undef, float %27, 0
  %29 = insertvalue { float, float } %28, float %"'ipse10", 1
  ret { float, float } %29
}

ERROR: LLVM error: function failed verification (4)
```